ampd/src/broadcaster_v2/mod.rs

```diff
@@ -182,40 +261,83 @@ where
                 )
             })
             .await
+            .inspect(|res| {
+                info!(
+                    tx_hash = res.txhash,
+                    msg_count, "successfully broadcasted tx"
+                );
+            })
     }
-}
 
-#[instrument]
-fn handle_tx_res(tx_hash: Result<String>, msgs: nonempty::Vec<msg_queue::QueueMsg>) {
-    Vec::from(msgs)
-        .into_iter()
-        .enumerate()
-        .for_each(|(i, msg)| {
-            match (msg.tx_res_callback, &tx_hash) {
-                (tx_res_callback, Ok(tx_hash)) => {
-                    let _ = tx_res_callback.send(Ok((tx_hash.clone(), i as u64)));
-                }
-                (tx_res_callback, Err(err)) => {
-                    let _ = tx_res_callback.send(Err(report!(err.current_context().clone())));
-                }
-            };
-        });
+    async fn estimate_fee(&mut self, batch_req: Any) -> Result<Fee> {
+        let gas = self
+            .broadcaster
+            .estimate_gas(vec![batch_req])
+            .await
+            .change_context(Error::EstimateGas)?;
+        let gas = gas as f64 * self.gas_adjustment;
+
+        Ok(Fee::from_amount_and_gas(
+            Coin::new(
+                cast(gas.mul(self.gas_price.amount).ceil()).ok_or(report!(Error::FeeAdjustment))?,
+                self.gas_price.denom.as_ref(),
+            )
+            .change_context(Error::FeeAdjustment)?,
+            cast::<f64, u64>(gas).ok_or(report!(Error::FeeAdjustment))?,
+        ))
+    }
+
+    #[instrument(skip(self))]
+    async fn handle_tx_res(
+        &self,
+        tx_hash: Result<String>,
+        msgs: nonempty::Vec<msg_queue::QueueMsg>,
+    ) -> Result<()> {
+        if let (Some(confirmer), Ok(tx_hash)) = (&self.tx_confirmer_client, &tx_hash) {
+            confirmer
+                .send(tx_hash.clone())
+                .await
+                .map_err(|_| Error::ConfirmTx(tx_hash.clone()))?;
```

Comment:
do we still need to throw the error out here, now that it's decoupled from the callback?

Reply:
what do you mean? the callback is still a few lines below? Besides I think its ok to throw the error if the confirmer receiver is closed instead of just ignoring it and send the unconfirmed `tx_hash` to the callback , or?